### PR TITLE
141-Refactor MaterialFactory

### LINKED ISSUE #141: Refactor MaterialFactory

### PR DESCRIPTION
**Pull request by uhetmaniuk**
Closes #141 

These modifications will simplify the creation of one executable.

### CODE REVIEW

**Review by japlews — CHANGES_REQUESTED** [review]
I would prefer if you didn't push this. NimbleSMExtras utilizes the BlockMaterialInterfaceFactory and MaterialFactory to inject types and behaviors into NimbleSM. Maybe we can have a broader discussion about the future of NimbleSMExtras @uhetmaniuk @rjones-sandia?

**Comment by nmm0** [comment]
> I would prefer if you didn't push this. NimbleSMExtras utilizes the BlockMaterialInterfaceFactory and MaterialFactory to inject types and behaviors into NimbleSM. Maybe we can have a broader discussion about the future of NimbleSMExtras @uhetmaniuk @rjones-sandia?

Is there a way to keep NimbleSMExtras working with the refactor? Maybe this would be a good topic to discuss at the next meeting

**Comment by japlews** [comment]
@nmm0 Yes, I think some changes might be required to this PR, but we could come to some consensus on the interface for Extras. It may be reasonable to set up some unit tests in NimbleSM which capture the interface use cases so that future changes don't break Nimble-as-a-library.

**Comment by uhetmaniuk** [comment]
@nmm0 @japlews @rjones-sandia 

Talking about it sounds good.

I chose the approach of defining an abstract class `nimble::MaterialFactoryBase` in order to avoid any `Kokkos` dependency. The classes `nimble::MaterialFactory` and `nimble_kokkos::MaterialFactory` are derived from the base class. But we can change that approach.

Another class that saw similar changes is `nimble::ModelData`: an abstract base class is defined `nimble::BaseModelData` and two derived classes `nimble::ModelData` and `nimble_kokkos::ModelData`.  This class hierarchy could also be modified if it conflicts with other features.

File: src/nimble_kokkos.h [hunk 1]
```diff
@@ -63,10 +63,7 @@ struct NimbleKokkosInitData {
 
 NimbleKokkosInitData NimbleKokkosInitializeAndGetInput(int argc, char* argv[]);
 
-void NimbleKokkosMain(std::shared_ptr<nimble_kokkos::MaterialFactory> material_factory,
```
Comment (japlews) [inline src/nimble_kokkos.h:66]: nimble_kokkos::MaterialFactory, nimble_kokkos::ContactInterface, nimble_kokkos::BlockMaterialInterfaceFactory, and nimble::Parser are all classes which are overriden in Extras; so some interface to create these polymorphic objects outside of NimbleSM and pass them in is needed

File: src/nimble_mpi.cc [hunk 2]
```diff
@@ -313,6 +315,21 @@ int NimbleMain(std::shared_ptr<MaterialFactoryType> material_factory,
   int contact_force_field_id = model_data.AllocateNodeData(nimble::VECTOR, "contact_force", num_nodes);
   int skin_node_field_id = model_data.AllocateNodeData(nimble::SCALAR, "skin_node", num_nodes);
 
+  //
```
Comment (japlews) [inline src/nimble_mpi.cc:318]: Same comment here: the type of the factory created needs to be prescribed through some kind of (preferably abstract) interface for compatibility with Extras.

**Comment by uhetmaniuk** [comment]
Closing this pull request as it is superseded with other updates.

**Pull request closed without merging**